src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -533,23 +533,29 @@ public void OnCompleted(Func<object, Task> callback, object state)
 
         public void Flush()
         {
-            ProduceStartAndFireOnStarting().GetAwaiter().GetResult();
+            InitializeResponse(0).GetAwaiter().GetResult();
             Output.Flush();
         }
 
         public async Task FlushAsync(CancellationToken cancellationToken)
         {
-            await ProduceStartAndFireOnStarting();
+            await InitializeResponse(0);
             await Output.FlushAsync(cancellationToken);
         }
 
         public void Write(ArraySegment<byte> data)
         {
-            // For the first write, ensure headers are flushed if Write(Chunked)isn't called.
+            // For the first write, ensure headers are flushed if Write(Chunked) isn't called.
             var firstWrite = !HasResponseStarted;
 
-            VerifyAndUpdateWrite(data.Count);
-            ProduceStartAndFireOnStarting().GetAwaiter().GetResult();
+            if (firstWrite)
+            {
+                InitializeResponse(data.Count).GetAwaiter().GetResult();
+            }
+            else
+            {
+                VerifyAndUpdateWrite(data.Count);
```

Comment:
Can we change InitializeResponse to not call VerifyAndUpdateWrite, and just call VerifyAndUpdateWrite afterwards?